### ISSUE TITLE
[spv-in] SPIRV Parsing Failures for BVE-Reborn and Rend3

### ISSUE THREAD

**Issue by cwfitzgerald** [gfx-rs/naga#265]
Here all the current problems preventing bve-reborn and rend3 from getting full shader validation. I have updated both to the latest wgpu master, so I may have issues that were already fixed but not yet pulled all the way through.

### Rend3:

This is a bindless based renderer, so uses more advanced features.

- `UnsupportedInstruction(ExecutionMode, String)`
  [Shader Playground](http://shader-playground.timjones.io/e1732368c06338ab32096f943c7e8dc7)
  [SPIR-V](https://github.com/gfx-rs/naga/files/5511895/pre_cull.zip)
- `UnsupportedInstruction(EntryPoint, String)`
  [Shader Playground](http://shader-playground.timjones.io/096642565325c39bacb16b7984803edc)
  [SPIR-V](https://github.com/gfx-rs/naga/files/5511913/blit.zip)
- `UnsupportedCapability(ShaderNonUniform)`
  [Shader Playground](http://shader-playground.timjones.io/c1d45c1960ca7dd40234eaaae3caf0ad)
  [SPIR-V](https://github.com/gfx-rs/naga/files/5511941/depth.zip)
  This also has `RuntimeDescriptorArray` and `SampledImageArrayNonUniformIndexing` which are likely unsupported.

### BVE-Reborn

More traditional renderer:

- `UnsupportedInstruction(ExecutionMode, String)`
  [Shader Playground](http://shader-playground.timjones.io/12bf63f7b226a52271c3c20c3d67fd81)
  [SPIR-V](https://github.com/gfx-rs/naga/files/5512112/opaque.zip)

### Other Info

All other errors I believe to be duplicates of the above.

**Comment by Systemcluster** [comment]
Two of these shaders (`pre_cull.spv` and `opaque.spv`) currently generate the following error:

```
...
[2021-02-14T17:54:15Z DEBUG naga::front::spv]           Return [1]
[2021-02-14T17:54:15Z DEBUG naga::front::spv::function] FunctionEnd
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: InvalidId(24)', examples\convert.rs:95:64
```

When `lookup_function.lookup(fun_id)` is called from `patch_function_calls` it's empty, so the lookup fails.
https://github.com/gfx-rs/naga/blob/7a124550d4f75a7089d1fe9d415a87e042fd88a8/src/front/spv/mod.rs#L1615

`patch_function_calls` is only called here:
https://github.com/gfx-rs/naga/blob/7a124550d4f75a7089d1fe9d415a87e042fd88a8/src/front/spv/function.rs#L149

Should it be populated before that line?

`blit.spv` parses now 🎉

**Comment by kvark** [comment]
Were you running Naga from latest master, or are these errors from `gfx-10` tag that is used by `wgpu` at the moment?
I think the issue about function calls is fixed now in #458 and will be included in `gfx-11` tag/train when it ships.

**Comment by Systemcluster** [comment]
I'm running the latest master, so #458 included. Before #458 it fails inside `fun.fill_global_use` instead.

**Comment by kvark** [comment]
Do you by any chance have forward function calling there? I.e. the function being called goes after the function that is the caller?

**Comment by kvark** [comment]
For now, we require that any functions called from something are defined prior to something. So we don't support forward calls or recursive calls. The spirv frontend should gracefully fail otherwise, and it kinda does, just without a proper error message.

**Comment by Systemcluster** [comment]
I don't think so, no. The corresponding GLSL for `pre_cull.spv` is [this](https://shader-playground.timjones.io/e1732368c06338ab32096f943c7e8dc7). It's the `bool visible = frustum_contains_sphere(uniforms.frustum, vec4(mesh_center, mesh_radius));` line that seems to cause it there.

**Comment by kvark** [comment]
Interesting. So it is defined prior in GLSL, but SPIR-V puts is behind :/
That's unfortunate, and I haven't seen this behavior before. Running on Dota2 shader set didn't produce anything like this.
We need some logic to anticipate this and reorder the functions.
You can probably work around it by running `spirv-opt` on the shaders on your side.

**Comment by kvark** [comment]
Filed #573 for this problem
